packages/builders/src/components/Component.ts

```diff
@@ -1,12 +1,31 @@
-import type { APIMessageComponent, ComponentType } from 'discord-api-types/v9';
 import type { JSONEncodable } from '../util/jsonEncodable';
+import type { APIBaseMessageComponent, APIMessageComponent, ComponentType } from 'discord-api-types/v9';
 
 /**
  * Represents a discord component
  */
-export interface Component extends JSONEncodable<APIMessageComponent> {
+export abstract class Component<
+	DataType extends APIBaseMessageComponent<ComponentType> = APIBaseMessageComponent<ComponentType>,
+> implements JSONEncodable<APIMessageComponent>
+{
+	/**
+	 * The API data associated with this component
+	 */
+	protected data: DataType;
```

Comment:
You don't set this.data again right, just its properties? If so, mark it as readonly

```suggestion
	protected readonly data: DataType;
```